src/main/java/org/opensearch/neuralsearch/processor/NeuralSparseTwoPhaseProcessor.java

```diff
@@ -0,0 +1,230 @@
+            }
+
+            return new NeuralSparseTwoPhaseProcessor(tag, description, ignoreFailure, enabled, ratio, window_expansion, max_window_size);
+        }
+    }
+
+    private QueryBuilder getNestedQueryBuilderFromNeuralSparseQueryBuilderMap(
+        final Multimap<NeuralSparseQueryBuilder, Float> queryBuilderFloatMap
+    ) {
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        queryBuilderFloatMap.asMap().forEach((neuralSparseQueryBuilder, boosts) -> {
+            float reduceBoost = boosts.stream().reduce(0.0f, Float::sum);
+            final float finalBoost = neuralSparseQueryBuilder.boost() * reduceBoost;
+            boolQueryBuilder.should(neuralSparseQueryBuilder.boost(finalBoost));
+        });
+        return boolQueryBuilder;
+    }
+
+    private float getOriginQueryWeightAfterRescore(final SearchSourceBuilder searchSourceBuilder) {
+        if (searchSourceBuilder.rescores() == null) return 1.0f;
+        return searchSourceBuilder.rescores()
+            .stream()
+            .map(rescorerBuilder -> ((QueryRescorerBuilder) rescorerBuilder).getQueryWeight())
+            .reduce(1.0f, (a, b) -> a * b);
+    }
+
+    private void collectNeuralSparseQueryBuilder(
+        final QueryBuilder queryBuilder,
+        final Multimap<NeuralSparseQueryBuilder, Float> neuralSparseQueryBuilderFloatMap,
+        float baseBoost
+    ) {
+        if (queryBuilder instanceof BoolQueryBuilder) {
+            collectNeuralSparseQueryBuilder((BoolQueryBuilder) queryBuilder, neuralSparseQueryBuilderFloatMap, baseBoost);
+        } else if (queryBuilder instanceof NeuralSparseQueryBuilder) {
+            collectNeuralSparseQueryBuilder((NeuralSparseQueryBuilder) queryBuilder, neuralSparseQueryBuilderFloatMap, baseBoost);
+        }
+    }
+
+    private void collectNeuralSparseQueryBuilder(
+        final BoolQueryBuilder queryBuilder,
+        final Multimap<NeuralSparseQueryBuilder, Float> neuralSparseQueryBuilderFloatMap,
+        float baseBoost
+    ) {
+        baseBoost *= queryBuilder.boost();
+        for (QueryBuilder subQuery : queryBuilder.should()) {
+            if (subQuery instanceof BoolQueryBuilder) {
+                collectNeuralSparseQueryBuilder(subQuery, neuralSparseQueryBuilderFloatMap, baseBoost);
+            } else if (subQuery instanceof NeuralSparseQueryBuilder) {
+                collectNeuralSparseQueryBuilder((NeuralSparseQueryBuilder) subQuery, neuralSparseQueryBuilderFloatMap, baseBoost);
+            }
+        }
+    }
+
+    private void collectNeuralSparseQueryBuilder(
+        final NeuralSparseQueryBuilder queryBuilder,
+        final Multimap<NeuralSparseQueryBuilder, Float> neuralSparseQueryBuilderFloatMap,
+        float baseBoost
+    ) {
+        baseBoost *= queryBuilder.boost();
+        neuralSparseQueryBuilderFloatMap.put(queryBuilder.getCopyNeuralSparseQueryBuilderForTwoPhase(ratio), baseBoost);
```

Comment:
can we rewrite it to a form when input arguments are not mutated inside the method that is called recursively. We can return pair of objects to the caller.

Reply:
Changed.